overlord/ifacestate/handlers.go

```diff
@@ -1532,6 +1534,121 @@ func (m *InterfaceManager) doHotplugDisconnect(task *state.Task, _ *tomb.Tomb) e
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	systemSnap, err := systemSnapInfo(st)
+	if err != nil {
+		return fmt.Errorf("core snap not available")
+	}
+
+	ifaceName, hotplugKey, err := getHotplugAttrs(task)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get hotplug task attributes: %s", err)
+	}
+
+	var slotSpec hotplug.RequestedSlotSpec
+	if err := task.Get("slot-spec", &slotSpec); err != nil {
+		return fmt.Errorf("internal error: cannot get hotplug slot specification from task attributes: %s", err)
+	}
+	var devinfo hotplug.HotplugDeviceInfo
+	if err := task.Get("device-info", &devinfo); err != nil {
+		return fmt.Errorf("internal error: cannot get hotplug device info from task attributes: %s", err)
+	}
+
+	stateSlots, err := getHotplugSlots(st)
+	if err != nil {
+		return fmt.Errorf("internal error obtaining hotplug slots: %v", err.Error())
+	}
+
+	iface := m.repo.Interface(ifaceName)
+	if iface == nil {
+		return fmt.Errorf("internal error: cannot find interface %s", ifaceName)
+	}
+
+	slot := findHotplugSlot(stateSlots, ifaceName, hotplugKey)
+
+	// if we know this slot already, restore / update it.
+	if slot != nil {
+		if slot.HotplugGone {
+			// hotplugGone means the device was unplugged, so its disconnect hooks were run and can now
+			// simply recreate the slot with potentially new attributes, and old connections will be re-created
+			newSlot := &snap.SlotInfo{
+				Name:       slot.Name,
+				Label:      slotSpec.Label,
+				Snap:       systemSnap,
+				Interface:  ifaceName,
+				Attrs:      slotSpec.Attrs,
+				HotplugKey: hotplugKey,
+			}
+			if iface, ok := iface.(interfaces.SlotSanitizer); ok {
+				if err := iface.BeforePrepareSlot(newSlot); err != nil {
+					return fmt.Errorf("cannot sanitize hotplug slot %q for interface %s: %s", newSlot.Name, newSlot.Interface, err)
+				}
+			}
+
+			if err := m.repo.AddSlot(newSlot); err != nil {
+				return fmt.Errorf("cannot restore hotplug slot %q for interface %s: %s", slot.Name, slot.Interface, err)
+			}
+			slot.HotplugGone = false
+			slot.StaticAttrs = slotSpec.Attrs
+			stateSlots[slot.Name] = slot
+			setHotplugSlots(st, stateSlots)
+
+			return nil
+		}
+
+		// else - not gone, restored already by reloadConnections, but may need updating.
+		if !reflect.DeepEqual(slotSpec.Attrs, slot.StaticAttrs) {
+			ts := updateDevice(st, iface.Name(), hotplugKey, slotSpec.Attrs)
+			snapstate.InjectTasks(task, ts)
+			st.EnsureBefore(0)
+		} // else - nothing to do
+		return nil
+	}
+
+	// New slot. Determine the name: use name provided by slot spec if set, otherwise use auto-generated name
+	proposedName := slotSpec.Name
+	if proposedName == "" {
+		proposedName = suggestedSlotName(&devinfo, ifaceName)
+	}
+	proposedName = ensureUniqueName(proposedName, func(name string) bool {
+		if slot, ok := stateSlots[name]; ok {
+			return slot.HotplugKey == hotplugKey
+		}
+		return m.repo.Slot(systemSnap.InstanceName(), name) == nil
```

Comment:
because we use sometimes  slots in the repo and sometimes  hotplug-slots in state we really need to make sure they are sync. Maybe we should have a set of highlevel helpers that always do both taking together a repo and state.

Reply:
Extracted into a helper, will address naming rules in a followup.

Reply:
I've added a storeHotplugSlot helper for the case you commented on above, this one was indeed a clear win. I don't see more cases for such helper in the already landed code though, the repo only reflects state when we (re)add a slot, e.g. hotplugGone flag is true, otherwise they are not in sync.